src/thread_task.rs

```diff
@@ -813,869 +813,746 @@ unsafe fn delayed_fg_task(c: *const Rav1dContext, ttd: *mut TaskThreadData) {
-                                        as c_long
-                                        as c_int;
-                                    let ts_0: *mut Rav1dTileState = &mut *((*f).ts)
-                                        .offset(tile_idx as isize)
-                                        as *mut Rav1dTileState;
-                                    (*tc).ts = ts_0;
-                                    (*tc).by = sby << (*f).sb_shift;
-                                    let uses_2pass = ((*c).n_fc > 1 as c_uint) as c_int;
-                                    (*tc).frame_thread.pass = if uses_2pass == 0 {
-                                        0 as c_int
-                                    } else {
-                                        1 as c_int
-                                            + ((*t).type_0 as c_uint
-                                                == RAV1D_TASK_TYPE_TILE_RECONSTRUCTION as c_int
-                                                    as c_uint)
-                                                as c_int
-                                    };
-                                    if error_0 == 0 {
-                                        error_0 = match rav1d_decode_tile_sbrow(&mut *tc) {
-                                            Ok(()) => 0,
-                                            Err(()) => 1,
-                                        };
-                                    }
-                                    let progress = if error_0 != 0 { TILE_ERROR } else { 1 + sby };
-                                    ::core::intrinsics::atomic_or_seqcst(
+                (*ttd).cur = ((*ttd).cur).wrapping_add(1);
+            }
+        }
+        if !found {
+            if reset_task_cur(c, ttd, u32::MAX) != 0 {
+                continue;
+            }
+            if merge_pending(c) != 0 {
+                continue;
+            }
+            park(tc, ttd);
+            continue;
+        }
+
+        // found:
+        // remove t from list
+        if !prev_t.is_null() {
+            (*prev_t).next = (*t).next;
+        } else {
+            (*f).task_thread.task_head = (*t).next;
+        }
+        if ((*t).next).is_null() {
+            (*f).task_thread.task_tail = prev_t;
+        }
+        if (*t).type_0 as c_uint > RAV1D_TASK_TYPE_INIT_CDF as c_int as c_uint
+            && ((*f).task_thread.task_head).is_null()
+        {
+            (*ttd).cur = ((*ttd).cur).wrapping_add(1);
+        }
+        (*t).next = 0 as *mut Rav1dTask;
+        // we don't need to check cond_signaled here, since we found a task
+        // after the last signal so we want to re-signal the next waiting thread
+        // and again won't need to signal after that
+        ::core::intrinsics::atomic_store_seqcst(&mut (*ttd).cond_signaled, 1 as c_int);
+        pthread_cond_signal(&mut (*ttd).cond);
+        pthread_mutex_unlock(&mut (*ttd).lock);
```

Comment:
```suggestion
        '_found: {
            // remove t from list
            if !prev_t.is_null() {
                (*prev_t).next = (*t).next;
            } else {
                (*f).task_thread.task_head = (*t).next;
            }
            if ((*t).next).is_null() {
                (*f).task_thread.task_tail = prev_t;
            }
            if (*t).type_0 as c_uint > RAV1D_TASK_TYPE_INIT_CDF as c_int as c_uint
                && ((*f).task_thread.task_head).is_null()
            {
                (*ttd).cur = ((*ttd).cur).wrapping_add(1);
            }
            (*t).next = 0 as *mut Rav1dTask;
            // we don't need to check cond_signaled here, since we found a task
            // after the last signal so we want to re-signal the next waiting thread
            // and again won't need to signal after that
            ::core::intrinsics::atomic_store_seqcst(&mut (*ttd).cond_signaled, 1 as c_int);
            pthread_cond_signal(&mut (*ttd).cond);
            pthread_mutex_unlock(&mut (*ttd).lock);
        }
```

Writing `goto` labels as labeled Rust blocks helps with clarity I think and matching it with the C (vs. just a comment).

Reply:
I'd prefer not to add labeled blocks unless we're actually using them. A comment is just as readable but doesn't require indenting the inner code. For the 'next block where we have shared code, that's great, but here we wouldn't actually break out of the block early.